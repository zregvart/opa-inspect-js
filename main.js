import * as fs from 'node:fs';
import * as polyfill from './polyfill.js';
import * as go from './wasm_exec.js';

const inspect = (f, m = null) => {
  const goruntime = new Go();
  const exitListener = code => {
    if (code === 0 && !goruntime.exited) {
      goruntime._pendingEvent = { id: 0 };
      goruntime._resume();
    }
  }
  process.on('exit', exitListener);
  // main.that will assign functions to this insance
  const that = {}
  // pass the reference so it's accessible in main.that
  goruntime.that = that

  return new Promise((res, rej) => {
    WebAssembly.instantiate(
      fs.readFileSync(new URL('./inspect.wasm', import.meta.url)),
      goruntime.importObject
    )
    .then(result => {
      goruntime.run(result.instance);
    })
    .then(() => {
      that.read = (path) => fs.readFileSync(path)

      const p = m == null ? that.inspect(f) : that.inspect(f, m);
      p.then(val => {
        res(JSON.parse(val));
      })
      .catch(rej)
    })
    .catch(rej)
    .finally(() => {
      that.finish && that.finish()
      process.removeListener('exit', exitListener)
    });
  });
}

export { inspect }
