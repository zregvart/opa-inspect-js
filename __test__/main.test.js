import * as fs from 'node:fs';
import * as events from 'node:events';
import * as url from 'node:url';
import * as vfs from 'vinyl-fs';
import Vinyl from 'vinyl';
import * as opa from '../main';

testFile = (f) => url.fileURLToPath(new url.URL(f, import.meta.url))

describe("nodejs", () => {
    beforeAll(() => {
        events.EventEmitter.defaultMaxListeners = 10;
    })

    afterAll(() => {
        expect(process.listeners('exit')).toHaveLength(0);
    })

    test('inspects rego files', async () => {
        const rego = fs.readFileSync(testFile('./example.rego')).toString();
        const json = await opa.inspect('example.rego', rego);
        expect(json).toMatchSnapshot();
    });

    test('inspects rego files second time', async () => {
        const rego = fs.readFileSync(testFile('./example.rego')).toString();
        const json = await opa.inspect('example.rego', rego);
        expect(json).toMatchSnapshot();
    });

    test('reports error when parsing', async () => {
        await expect(opa.inspect('example.rego', "bogus")).rejects.toEqual("1 error occurred: example.rego:1: rego_parse_error: package expected");
    });

    test('inspects rego files read from the filesystem', async () => {
        const json = await opa.inspect(testFile('./example.rego'));
        expect(json).toHaveLength(1);
        expect(json[0]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example\.rego$/),
            }
        });
    });

    test('inspects multiple rego files', async () => {
        const json = await opa.inspect([testFile('./example.rego'), testFile('./example2.rego')]);
        expect(json).toHaveLength(2);
        expect(json[0]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example\.rego$/),
            }
        });
        expect(json[1]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example2\.rego$/),
            }
        });
    });

    test('inspects vinyl streams', async () => {
        const json = await opa.inspect(vfs.src('__test__/*.rego'));

        expect(json).toHaveLength(2);
        expect(json[0]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example\.rego$/),
            }
        });
        expect(json[1]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example2\.rego$/),
            }
        });
    });

    test('multiple vinyl files', async () => {
        const base = url.fileURLToPath(import.meta.url)
        const example = new Vinyl({
            cwd: '/',
            base: base,
            path: `/${base}/example.rego`,
            contents: fs.readFileSync(testFile('./example.rego', import.meta.url))
        });

        const example2 = new Vinyl({
            cwd: '/',
            base: base,
            path: `/${base}/example2.rego`,
            contents: fs.readFileSync(testFile('./example2.rego'))
        });

        const json = await opa.inspect([example, example2]);

        expect(json).toHaveLength(2);
        expect(json[0]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example\.rego$/),
            }
        });
        expect(json[1]).toMatchSnapshot({
            "location": {
                "file": expect.stringMatching(/\/example2\.rego$/),
            }
        });
    });

    test.concurrent.each([...Array(25).keys()])('functions concurrently (%#)', async () => {
        const rego = fs.readFileSync(new url.URL('./example.rego', import.meta.url)).toString();
        await expect(opa.inspect('example.rego', rego)).resolves.toHaveProperty('[0].annotations');
    });

    test('package metadata', async () => {
        const json = await opa.inspect('example.rego', `#
# METADATA
# title: title
# description: description
package example`);
        expect(json).toMatchSnapshot();
    })
});
