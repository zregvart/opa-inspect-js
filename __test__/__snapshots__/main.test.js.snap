// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodejs inspects multiple rego files 1`] = `
{
  "annotations": {
    "custom": {
      "failure_msg": "Task '%s' does not contain a bundle reference",
      "short_name": "disallowed_task_reference",
    },
    "description": "Check for existence of a task bundle. Enforcing this rule will
fail the contract if the task is not called from a bundle.",
    "scope": "rule",
    "title": "Task bundle was not used or is not defined",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example\\\\\\.rego\\$/,
    "row": 12,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;

exports[`nodejs inspects multiple rego files 2`] = `
{
  "annotations": {
    "custom": {
      "short_name": "example2",
    },
    "description": "Second example",
    "scope": "rule",
    "title": "Example 2",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example2\\\\\\.rego\\$/,
    "row": 8,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example2",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;

exports[`nodejs inspects rego files 1`] = `
[
  {
    "annotations": {
      "custom": {
        "failure_msg": "Task '%s' does not contain a bundle reference",
        "short_name": "disallowed_task_reference",
      },
      "description": "Check for existence of a task bundle. Enforcing this rule will
fail the contract if the task is not called from a bundle.",
      "scope": "rule",
      "title": "Task bundle was not used or is not defined",
    },
    "location": {
      "col": 1,
      "file": "example.rego",
      "row": 12,
    },
    "path": [
      {
        "type": "var",
        "value": "data",
      },
      {
        "type": "string",
        "value": "example",
      },
      {
        "type": "string",
        "value": "deny",
      },
    ],
  },
]
`;

exports[`nodejs inspects rego files read from the filesystem 1`] = `
{
  "annotations": {
    "custom": {
      "failure_msg": "Task '%s' does not contain a bundle reference",
      "short_name": "disallowed_task_reference",
    },
    "description": "Check for existence of a task bundle. Enforcing this rule will
fail the contract if the task is not called from a bundle.",
    "scope": "rule",
    "title": "Task bundle was not used or is not defined",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example\\\\\\.rego\\$/,
    "row": 12,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;

exports[`nodejs inspects rego files second time 1`] = `
[
  {
    "annotations": {
      "custom": {
        "failure_msg": "Task '%s' does not contain a bundle reference",
        "short_name": "disallowed_task_reference",
      },
      "description": "Check for existence of a task bundle. Enforcing this rule will
fail the contract if the task is not called from a bundle.",
      "scope": "rule",
      "title": "Task bundle was not used or is not defined",
    },
    "location": {
      "col": 1,
      "file": "example.rego",
      "row": 12,
    },
    "path": [
      {
        "type": "var",
        "value": "data",
      },
      {
        "type": "string",
        "value": "example",
      },
      {
        "type": "string",
        "value": "deny",
      },
    ],
  },
]
`;

exports[`nodejs inspects vinyl streams 1`] = `
{
  "annotations": {
    "custom": {
      "failure_msg": "Task '%s' does not contain a bundle reference",
      "short_name": "disallowed_task_reference",
    },
    "description": "Check for existence of a task bundle. Enforcing this rule will
fail the contract if the task is not called from a bundle.",
    "scope": "rule",
    "title": "Task bundle was not used or is not defined",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example\\\\\\.rego\\$/,
    "row": 12,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;

exports[`nodejs inspects vinyl streams 2`] = `
{
  "annotations": {
    "custom": {
      "short_name": "example2",
    },
    "description": "Second example",
    "scope": "rule",
    "title": "Example 2",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example2\\\\\\.rego\\$/,
    "row": 8,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example2",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;

exports[`nodejs multiple vinyl files 1`] = `
{
  "annotations": {
    "custom": {
      "failure_msg": "Task '%s' does not contain a bundle reference",
      "short_name": "disallowed_task_reference",
    },
    "description": "Check for existence of a task bundle. Enforcing this rule will
fail the contract if the task is not called from a bundle.",
    "scope": "rule",
    "title": "Task bundle was not used or is not defined",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example\\\\\\.rego\\$/,
    "row": 12,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;

exports[`nodejs multiple vinyl files 2`] = `
{
  "annotations": {
    "custom": {
      "short_name": "example2",
    },
    "description": "Second example",
    "scope": "rule",
    "title": "Example 2",
  },
  "location": {
    "col": 1,
    "file": StringMatching /\\\\/example2\\\\\\.rego\\$/,
    "row": 8,
  },
  "path": [
    {
      "type": "var",
      "value": "data",
    },
    {
      "type": "string",
      "value": "example2",
    },
    {
      "type": "string",
      "value": "deny",
    },
  ],
}
`;
